slang-checks/src/main/java/com/sonarsource/slang/checks/DuplicatedFunctionImplementationCheck.java

```diff
@@ -0,0 +1,108 @@
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import org.sonar.check.Rule;
+
+import static com.sonarsource.slang.utils.SyntacticEquivalence.areEquivalent;
+
+@Rule(key = "S4144")
+public class DuplicatedFunctionImplementationCheck implements SlangCheck {
+
+  private static final String MESSAGE = "Update this function so that its implementation is not identical to \"%s\" on line %s.";
+  private static final String MESSAGE_NO_NAME = "Update this function so that its implementation is not identical to the one on line %s.";
+
+  @Override
+  public void initialize(InitContext init) {
+    init.register(TopLevelTree.class, (ctx, tree) -> {
+      List<FunctionDeclarationTree> functionDeclarations = tree.descendants()
+        .filter(FunctionDeclarationTree.class::isInstance)
+        .map(FunctionDeclarationTree.class::cast)
+        .collect(Collectors.toList());
+
+      check(ctx, functionDeclarations);
+    });
+  }
+
+  private static void check(CheckContext ctx, List<FunctionDeclarationTree> functionDeclarations) {
+    Set<FunctionDeclarationTree> reportedDuplicates = new HashSet<>();
+    IntStream.range(0, functionDeclarations.size()).forEach(i -> {
+      FunctionDeclarationTree original = functionDeclarations.get(i);
+      functionDeclarations.stream()
+        .skip(i + 1L)
+        .filter(f -> !reportedDuplicates.contains(f))
+        .filter(DuplicatedFunctionImplementationCheck::hasMinimumLineNumber)
+        .filter(f -> areDuplicatedImplementation(original, f))
+        .forEach(duplicate -> {
+          reportDuplicate(ctx, original, duplicate);
+          reportedDuplicates.add(duplicate);
+        });
+    });
+
+  }
+
+  private static boolean hasMinimumLineNumber(FunctionDeclarationTree function) {
+    BlockTree functionBody = function.body();
+    if (functionBody == null) {
+      return false;
+    }
+    // TODO replace by > 2 LoC when available
+    TextRange textRange = functionBody.metaData().textRange();
+    return textRange.end().line() - textRange.start().line() >= 4;
+  }
+
+  private static boolean areDuplicatedImplementation(FunctionDeclarationTree original, FunctionDeclarationTree possibleDuplicate) {
+    return areEquivalent(original.formalParameters(), possibleDuplicate.formalParameters())
+      && areEquivalent(original.body(), possibleDuplicate.body(), true);
+  }
+
+  private static void reportDuplicate(CheckContext ctx, FunctionDeclarationTree original, FunctionDeclarationTree duplicate) {
+    IdentifierTree identifier = original.name();
+    String message;
+    Tree secondaryTree;
+    if (identifier != null) {
+      secondaryTree = identifier;
+      message = String.format(MESSAGE, identifier.name(), original.metaData().textRange().start().line());
+    } else {
+      secondaryTree = original;
+      message = String.format(MESSAGE_NO_NAME, original.metaData().textRange().start().line());
```

Comment:
Minor: you could factorize `original.metaData().textRange().start().line()` which is also used 3 lines above.